### ISSUE TITLE
"toggle text edit mode" button sometimes not visible

### ISSUE THREAD

**Issue by sspieker** [grafana/grafana#18037]
**What happened**:
When editing a panel with a datasource that supports text edit mode, sometimes the "toggle text edit mode" button (the pen) is not shown initially. Furthermore, adding another query _does_ render the button for the new query but still _not_ for the first one. The issue can be solved by navigation to another tab (for example "visualisation") and back.
Here's a screenshot:

![noEditPen](https://user-images.githubusercontent.com/11644606/60977179-4b157a80-a32f-11e9-88a1-81d46c7b572f.png)

**What you expected to happen**:
A datasource that supports the text edit mode should _always_ show the "toggle text edit mode" button.

**How to reproduce it (as minimally and precisely as possible)**:
Create some workload on your local machine (compile grafana for example) and enter a panel's edit mode. I've got a ~ 66% chance of this issue to happen.
_Edit: a small code change (see Comment [511332893](https://github.com/grafana/grafana/issues/18037#issuecomment-511332893)) got me to a 100% reproduction rate._

**Anything else we need to know?**:
The issue has been mentioned before: #15858 and #16393 . Obviously, increasing the timeout from 10ms to 100ms did not do the trick.

**Environment**:
- Grafana version: 6.2.5
- Data source type & version: custom datasource with "text edit mode" support
- OS Grafana is installed on: Ubuntu 16.04
- User OS & Browser: Ubuntu 16.04 & Chrome/Firefox latest versions
- Grafana plugins: just the custom datasource
- Others: -

**Comment by torkelo** [comment]
Strange, similar to https://github.com/grafana/grafana/issues/15858 

is it always on the first query?

**Comment by sspieker** [comment]
@torkelo , the majority of our dashboards/panels consists of only one query. So although I'm tempted to simply answer "yes", I'll make sure to do a couple of tests tomorrow and post the results here.

**Comment by sspieker** [comment]
@torkelo , it's not just the first query/button. Take a look at this screenshot:

![pen2](https://user-images.githubusercontent.com/11644606/61033532-fcfe8680-a3c3-11e9-907c-35b594b1b6ec.png)

**Comment by sspieker** [comment]
@torkelo , our users keep complaining about this issue, so I've decided to investigate further. I've managed to eliminate the issue by setting the following timeout from **100** to **2000**:

```
    // give angular time to compile
    setTimeout(() => {
      this.setState({ hasTextEditMode: !!this.angularScope.toggleEditorMode });
    }, 100);
```
> https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/panel_editor/QueryEditorRow.tsx#L132

I don't think setting the timeout this high is a good idea (unhappy users incoming again...), but I hope it points you in the right direction: it indeed _is_ a timing problem.

**Comment by sspieker** [comment]
This issue is still tagged with _needs more detail_. Is there anything that can be done to speed things up a bit? I'd be happy to provide more details, but my observations in the last comment seem pretty clear to me.

**Comment by kaydelaney** [comment]
@sspieker Hey, I had some free time so I took a look at the issue. Wasn't able to reproduce the original problem on my system, but suspect the PR I've put up should fix it. Let me know if it does or doesn't though! https://github.com/grafana/grafana/pull/18335

**Comment by sspieker** [comment]
@kaydelaney ,  I did a quick test with the latest Docker image (master) and everything worked as expected. Well done!
@torkelo Is this something that could be included in the 6.3 update as well? Really looking forward to use the fixed version. ;)
